### PR TITLE
Port additional VB runtime methods to dotnet/winforms

### PR DESCRIPTION
**Pull request by cston**
## Proposed changes

Continued port of Microsoft.VisualBasic from https://github.com/Microsoft/referencesource/tree/master/Microsoft.VisualBasic/runtime/msvbalib.

Changes to the code ported from referencesource:
- Cleaned up comments, and removed commented out code
- Removed attribute use {HostProtection,SecuritySafeCritical,SecurityPermission,ResourceConsumption,ResourceExposure}
- Removed permission demands

_See https://github.com/dotnet/winforms/pull/1675 for the initial port._

## Test methodology

Unit tests for a number of methods, primarily testing simple code paths since this is a port of existing code. Integration tests for methods that rely on System.Windows.Forms.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/1871)

### CODE REVIEW

**Comment by cston** [comment]
cc @jaredpar, @danmosemsft, @KathleenDollard

**Comment by cston** [comment]
@dotnet/dotnet-winforms, @ericstj, @333fred, please review.

File: src/Microsoft.VisualBasic/tests/IntegrationTests/Microsoft.VisualBasic.IntegrationTests/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBaseTests.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    public class WindowsFormsApplicationBaseTests
+    {
+        [Fact]
+        public void Run()
+        {
+            string exePath = TestHelpers.GetExePath("VisualBasicRuntimeTest");
+            var startInfo = new ProcessStartInfo { FileName = exePath, Arguments = "WindowsFormsApplicationBase.Run" };
+            var process = TestHelpers.StartProcess(startInfo);
+            Assert.False(process.HasExited);
+            TestHelpers.EndProcess(process, timeout: 1000);
```
Comment (333fred) [inline src/Microsoft.VisualBasic/tests/IntegrationTests/Microsoft.VisualBasic.IntegrationTests/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBaseTests.cs]: Should we assert that the process has existed after?

File: src/Microsoft.VisualBasic/tests/UnitTests/Microsoft/VisualBasic/MyServices/ClipboardProxyTests.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            Assert.Equal(System.Windows.Forms.Clipboard.ContainsImage(), clipboard.ContainsImage());
+            Assert.Equal(System.Windows.Forms.Clipboard.GetImage(), clipboard.GetImage());
+            clipboard.SetImage(image);
+        }
+
+        [StaFact]
+        public void Audio()
+        {
+            var clipboard = (new Computer()).Clipboard;
+            Assert.Equal(System.Windows.Forms.Clipboard.ContainsAudio(), clipboard.ContainsAudio());
```
Comment (333fred) [inline src/Microsoft.VisualBasic/tests/UnitTests/Microsoft/VisualBasic/MyServices/ClipboardProxyTests.cs]: Today in new features that I've learned Windows has

File: src/Microsoft.VisualBasic/tests/UnitTests/Microsoft/VisualBasic/MyServices/FileSystemProxyTests.cs [hunk 3]
```diff
@@ -5,6 +5,7 @@
 using System;
 using System.Collections.Generic;
 using Microsoft.VisualBasic.FileIO;
+using Microsoft.VisualBasic.Tests;
```
Comment (333fred) [inline src/Microsoft.VisualBasic/tests/UnitTests/Microsoft/VisualBasic/MyServices/FileSystemProxyTests.cs]: Is this addition necessary?

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/ApplicationBase.vb [hunk 4]
```diff
@@ -39,6 +39,21 @@ Namespace Microsoft.VisualBasic.ApplicationServices
             Return VariableValue
         End Function
 
+        ''' <summary>
+        ''' Provides access to logging capability.
+        ''' </summary>
+        ''' <value>Returns a Microsoft.VisualBasic.Windows.Log object used for logging to OS log, debug window
+        ''' and a delimited text file or xml log.</value>
+        ''' <remarks></remarks>
```
Comment (333fred) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/ApplicationBase.vb]: Consider removing empty remarks block.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.vb [hunk 5]
```diff
@@ -84,5 +105,740 @@ Namespace Microsoft.VisualBasic.ApplicationServices
...
+        ''' <summary>
+        ''' Entry point to kick off the VB Startup/Shutdown Application model
+        ''' </summary>
+        ''' <param name="commandLine">The command line from Main()</param>
+        <SecuritySafeCritical()>
+        Public Sub Run(ByVal commandLine As String())
+            'Prime the command line args with what we receive from Main() so that Click-Once windows apps don't have to do a System.Environment call which would require permissions.
+            MyBase.InternalCommandLine = New System.Collections.ObjectModel.ReadOnlyCollection(Of String)(commandLine)
+
+            ' .NET Framework implementation uses System.Runtime.Remoting when  Me.IsSingleInstance
```
Comment (333fred) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.vb]: >   [](start = 77, length = 2)

Nit: extra space

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/Devices/ComputerInfo.vb [hunk 6]
```diff
@@ -82,6 +83,34 @@ Namespace Microsoft.VisualBasic.Devices
...
 
+        ''' <summary>
+        ''' Gets the full operating system name. This method requires full trust and WMI installed.
+        ''' </summary>
+        ''' <value>A string contains the operating system name.</value>
+        ''' <exception cref="System.Security.SecurityException">If the immediate caller does not have full trust.</exception>
+        ''' <exception cref="System.InvalidOperationException">If we cannot obtain the query object from WMI.</exception>
+        ''' <remarks>Since this property depends on WMI, we have OSPlatform property that does not require WMI.</remarks>
+        Public ReadOnly Property OSFullName() As String
+            Get
```
Comment (333fred) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/Devices/ComputerInfo.vb]: This had a `SecuritySafeCritical` in the referencesource, do we need one here as well?

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/Devices/ComputerInfo.vb [hunk 7]
```diff
@@ -179,6 +208,45 @@ Namespace Microsoft.VisualBasic.Devices
...
         End Property
 
+        ''' <summary>
+        ''' Get the management object used in WMI to query for the operating system name.
+        ''' </summary>
+        ''' <value>A ManagementBaseObject represents the result of "Win32_OperatingSystem" query.</value>
+        ''' <exception cref="System.Security.SecurityException">If the immediate caller does not have full trust.</exception>
+        ''' <exception cref="System.InvalidOperationException">If we cannot obtain the query object from WMI.</exception>
+        Private ReadOnly Property OSManagementBaseObject() As ManagementBaseObject
+            Get
```
Comment (333fred) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/Devices/ComputerInfo.vb]: Same question

File: eng/Versions.props [hunk 8]
```diff
@@ -20,6 +20,8 @@
     <SystemCodeDomPackageVersion>5.0.0-alpha1.19502.5</SystemCodeDomPackageVersion>
     <SystemConfigurationConfigurationManagerPackageVersion>5.0.0-alpha1.19502.5</SystemConfigurationConfigurationManagerPackageVersion>
     <SystemDrawingCommonPackageVersion>5.0.0-alpha1.19502.5</SystemDrawingCommonPackageVersion>
+    <SystemManagementExtensionsPackageVersion>5.0.0-alpha1.19502.5</SystemManagementExtensionsPackageVersion>
+    <SystemIOPortsPackageVersion>5.0.0-alpha1.19502.5</SystemIOPortsPackageVersion>
```
Comment (ericstj) [inline eng/Versions.props]: These are interesting new growth to the shared framework.  You'll need to decide if these are implementation only, or if they are exposed for reference /cc @dagood @terrajobst

File: src/Microsoft.VisualBasic.Forms/src/Resources/Microsoft/VisualBasic/MyServices/Internal/ProgressDialog.resx [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  <resheader name="version">
+    <value>2.0</value>
+  </resheader>
+  <resheader name="reader">
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <resheader name="writer">
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+</root>
```
Comment (ericstj) [inline src/Microsoft.VisualBasic.Forms/src/Resources/Microsoft/VisualBasic/MyServices/Internal/ProgressDialog.resx]: This looks empty, was that intentional?

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/Helpers/FileSystemUtils.vb [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+Imports System.Globalization
+Imports System.Security
+Imports System.Text
+
+Imports ExUtils = Microsoft.VisualBasic.CompilerServices.ExceptionUtils
+
+Namespace Microsoft.VisualBasic.CompilerServices
+
+    ''' <summary>
+    '''  Internal utilities from Microsoft.VisualBasic.FileIO.FileSystem.
```
Comment (cston) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/Helpers/FileSystemUtils.vb]: Compare with https://github.com/microsoft/referencesource/blob/master/Microsoft.VisualBasic/runtime/msvbalib/FileIO/FileSystem.vb

**Comment by 333fred** [comment]
Done review pass (commit 1)

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.vb [hunk 11]
```diff
@@ -84,5 +105,740 @@ Namespace Microsoft.VisualBasic.ApplicationServices
...
+        Private m_FinishedOnInitilaize As Boolean 'Whether we have made it through the processing of OnInitialize 
+        Private m_NetworkAvailabilityEventHandlers As System.Collections.ArrayList
+        Private m_FirstInstanceSemaphore As System.Threading.EventWaitHandle 'Used to determine if we are the first instance or not and to prevent race conditions if we are launching subsequent instances.  Owned by the first instance to run
+        Private m_MessageReceivedSemaphore As System.Threading.EventWaitHandle 'Used to let us know when the first instance has been contacted
+        Private m_NetworkObject As Microsoft.VisualBasic.Devices.Network
+        Private m_MemoryMappedID As String 'global OS handles must have a unique ID
+        <SecurityCritical()>
+        Private m_FirstInstanceMemoryMappedFileHandle As SafeFileHandle 'Used to communicate the URL of the single instance between processes.  We hang on to it for the life of the process so that it is available to subsequent instances.  It's a SafeHandle so it'll be sure to release the handle when the app exits.
+        Private m_IsSingleInstance As Boolean 'whether this app runs using Word like instancing behavior
```
Comment (weltkante) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.vb]: There's a bunch of unused variables here considering the remoting-based implementation had to be dropped. Also I'm curious what the plan is for single instance support :-)

File: eng/Versions.props (hunk 8 above)
Comment (terrajobst) [reply eng/Versions.props]: Ouch. Indeed.

File: src/Microsoft.VisualBasic/tests/UnitTests/Microsoft/VisualBasic/MyServices/FileSystemProxyTests.cs (hunk 3 above)
Comment (cston) [reply src/Microsoft.VisualBasic/tests/UnitTests/Microsoft/VisualBasic/MyServices/FileSystemProxyTests.cs]: It was used for `FileCleanupTestBase` only. I've changed those type references to include the namespace.

---
In reply to: [331592564](https://github.com/dotnet/winforms/pull/1871#discussion_r331592564) [](ancestors = 331592564)

File: src/Microsoft.VisualBasic.Forms/src/Resources/Microsoft/VisualBasic/MyServices/Internal/ProgressDialog.resx (hunk 9 above)
Comment (cston) [reply src/Microsoft.VisualBasic.Forms/src/Resources/Microsoft/VisualBasic/MyServices/Internal/ProgressDialog.resx]: Yes. I'll add the resources for ProgressDialog and VBInputBox in a separate PR.

---
In reply to: [331604592](https://github.com/dotnet/winforms/pull/1871#discussion_r331604592) [](ancestors = 331604592)

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.vb (hunk 11 above)
Comment (cston) [reply src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.vb]: I've removed additional variables and logged [https://github.com/dotnet/winforms/issues/2056](https://github.com/dotnet/winforms/issues/2056) to consider support "single instance" behavior in the future.

---
In reply to: [331665470](https://github.com/dotnet/winforms/pull/1871#discussion_r331665470) [](ancestors = 331665470)

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/Devices/ComputerInfo.vb (hunk 6 above)
Comment (cston) [reply src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/Devices/ComputerInfo.vb]: `SecuritySafeCriticalAttribute` has no effect in .NET Core (see [https://docs.microsoft.com/en-us/dotnet/api/system.security.securitysafecriticalattribute?view=netcore-3.0#remarks](https://docs.microsoft.com/en-us/dotnet/api/system.security.securitysafecriticalattribute?view=netcore-3.0#remarks)).

---
In reply to: [331595800](https://github.com/dotnet/winforms/pull/1871#discussion_r331595800) [](ancestors = 331595800)

File: src/Microsoft.VisualBasic/tests/IntegrationTests/Microsoft.VisualBasic.IntegrationTests/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBaseTests.cs (hunk 1 above)
Comment (cston) [reply src/Microsoft.VisualBasic/tests/IntegrationTests/Microsoft.VisualBasic.IntegrationTests/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBaseTests.cs]: Added `Assert.True(process.HasExited);` after.

Removed `Assert.False(process.HasExited);` before to match [https://github.com/dotnet/winforms/pull/2050](https://github.com/dotnet/winforms/pull/2050).

---
In reply to: [331590488](https://github.com/dotnet/winforms/pull/1871#discussion_r331590488) [](ancestors = 331590488)

File: eng/Versions.props (hunk 8 above)
Comment (cston) [reply eng/Versions.props]: System.Management is used by `Microsoft.VisualBasic.Devices.ComputerInfo.OSFullName`.

System.IO.Ports is used by `Microsoft.VisualBasic.Devices.Ports`.


---
In reply to: [331603390](https://github.com/dotnet/winforms/pull/1871#discussion_r331603390) [](ancestors = 331603390)

Comment (cston) [reply eng/Versions.props]: Changed the implementation of `ComputerInfo.OSFullName` to use `System.Runtime.InteropServices.RuntimeInformation.OSDescription` and removed `Microsoft.VisualBasic.Devices.Ports`.

---
In reply to: [335213342](https://github.com/dotnet/winforms/pull/1871#discussion_r335213342) [](ancestors = 335213342,331603390)

**Comment by cston** [comment]
@dotnet/dotnet-winforms, @ericstj, @333fred, all feedback has been addressed. Please review, thanks.
